feat: forge deploy allocs

tynes opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**
Refactor op-e2e to use the same deploy script that is used in
production. There is only 1 way to deploy contracts now and
it goes through foundry. This removes the need to maintain
multiple sets of contracts deployments/devnets. It does
add a new tooling dep to the monorepo which is `geth`.
A top level `make` command is provided as well as some
docs in the `op-e2e` package on its usage.

The `devnetL1.json` deployconfig is the source of truth
for deploy configs in `op-e2e` now. In the long term,
we should delineate which config values are safe to alter
in memory and which are not. Some things are still coupled
together in weird ways, but I tried to add assertions where
I could to ensure that future people will get better error
messages.

To install geth, run the command:

```
make install-geth
```

To create the genesis state, run the command:

```
make devnet-allocs
```

This will run the forge deploy script against an instance of
`geth --dev` and then call `debug_dumpBlock` which will dump
the state into a format that we can inject at runtime into
the genesis of both `op-e2e` as well as the L1 of the docker
compose setup.

I tried to use `anvil` instead of `geth` but it does not
support this endpoint. It supports a different endpoint
that uses a compressed binary serialization of state,
it is easier to work with the JSON.

There are no more deterministic addresses on L1 because it is too much maintenance to maintain 2 deployment systems. Now the production deployment system is used in `op-e2e`

The addresses are stored in `.devnet/addresses.json` and I added an nginx server to serve them at `localhost:8080/addresses.json`. This is because they will be different every time and gives an easy way for the front end to access them.

Can confirm that the devnet blocktime + proposal time results in a usable local devnet, the L2 doesn't run away from the proposals

Alternative to https://github.com/ethereum-optimism/optimism/pull/6123

Closes CLI-4141

tynes committed:
op-e2e: major refactor

Refactor op-e2e to use the same deploy script that is used in
production. There is only 1 way to deploy contracts now and
it goes through foundry. This removes the need to maintain
multiple sets of contracts deployments/devnets. It does
add a new tooling dep to the monorepo which is `geth`.
A top level `make` command is provided as well as some
docs in the `op-e2e` package on its usage.

The `devnetL1.json` deployconfig is the source of truth
for deploy configs in `op-e2e` now. In the long term,
we should delineate which config values are safe to alter
in memory and which are not. Some things are still coupled
together in weird ways, but I tried to add assertions where
I could to ensure that future people will get better error
messages.

To install geth, run the command:

`` `
make install-geth
```

To create the genesis state, run the command:

```
make devnet-allocs
```

This will run the forge deploy script against an instance of
`geth --dev` and then call `debug_dumpBlock` which will dump
the state into a format that we can inject at runtime into
the genesis of both `op-e2e` as well as the L1 of the docker
compose setup.

I tried to use `anvil` instead of `geth` but it does not
support this endpoint. It supports a different endpoint
that uses a compressed binary serialization of state,
it is easier to work with the JSON.

tynes commented on op-e2e/actions/l2_proposer_test.go:
```diff
@@ -67,13 +78,32 @@ func TestProposer(gt *testing.T) {
+	idx, err := outputOracleContract.LatestBlockNumber(&bind.CallOpts{})
+	require.NoError(t, err)
+	require.Greater(t, int64(idx.Uint64()), int64(0), "output index must be greater than 0")
+	fmt.Println("==============", idx)
```
wip stuff here

tynes commented on op-e2e/e2eutils/secrets.go:
```diff
@@ -157,11 +148,40 @@ func (s *Secrets) Addresses() *Addresses {
 	}
 }
 
+func (s *Secrets) String() string {
```
This could prob be deleted

tynes commented on op-e2e/geth.go:
```diff
@@ -124,7 +124,7 @@ func initL1Geth(cfg *SystemConfig, genesis *core.Genesis, c clock.Clock, opts ..
 		HTTPModules: []string{"debug", "admin", "eth", "txpool", "net", "rpc", "web3", "personal", "engine"},
 	}
 
-	l1Node, l1Eth, err := createGethNode(false, nodeConfig, ethConfig, []*ecdsa.PrivateKey{cfg.Secrets.CliqueSigner}, opts...)
```
Was somehow L1 never mining blocks because this was not passed in?

tynes commented:
L1 genesis has pre merge config:

```
WARN [07-26|19:49:38.636] Served eth_getBlockByNumber              reqid=9 duration="161.419µs" err="'finalized' tag not supported on pre-merge network"
WARN [07-26|19:49:38.636] Served eth_getBlockByNumber              reqid=10 duration="166.15µs"  err="'safe' tag not supported on pre-merge network"
```

tynes commented on op-e2e/bridge_test.go:
```diff
@@ -8,20 +8,24 @@ import (
 func TestERC20BridgeDeposits(t *testing.T) {
 	InitParallel(t)
 
+	//runtime.Breakpoint()
```
dead code

tynes commented on op-e2e/setup.go:
```diff
@@ -454,6 +388,7 @@ func (cfg SystemConfig) Start(_opts ...SystemConfigOption) (*System, error) {
 		if name == "l1" {
 			continue
 		}
+		// In here someplace?
```
dead code

tynes commented on op-node/cmd/genesis/cmd.go:
```diff
@@ -55,7 +55,8 @@ var Subcommands = cli.Commands{
 			}
 
-			l1Genesis, err := genesis.BuildL1DeveloperGenesis(config)
+			// TODO: this needs new inputs now to be able to create an L1 genesis
```
Add an extra CLI arg for the path to the dump file

tynes commented on op-e2e/config/init.go:
```diff
@@ -0,0 +1,164 @@
+// ReadAllocs
+func ReadAllocs(filename string) (*state.Dump, error) {
+	_, err := os.Stat(filename)
+	if os.IsNotExist(err) {
```
`errors.Is(err, fs.ErrNotExist)`

tynes commented:
In the devnet with genesis contracts:

```
timeout 10m npx hardhat deposit-erc20 --network devnetL1
Using signer 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
An unexpected error occurred:

TypeError: Cannot read properties of undefined (reading 'l1')
    at SimpleTaskDefinition.action (/home/circleci/project/packages/sdk/tasks/deposit-erc20.ts:154:21)
```

tynes commented:
```
       	Error Trace:	/root/project/op-e2e/system_fpp_test.go:108
        	            				/root/project/op-e2e/system_fpp_test.go:31
        	Error:      	Received unexpected error:
        	            	Post "http://127.0.0.1:37781": context deadline exceeded
        	Test:       	TestVerifyL2OutputRootEmptyBlock
```

tynes commented on packages/sdk/src/interfaces/types.ts:
```diff
@@ -27,7 +27,7 @@ export enum L2ChainID {
   OPTIMISM_HARDHAT_LOCAL = 31337,
   OPTIMISM_HARDHAT_DEVNET = 17,
-  OPTIMISM_BEDROCK_LOCAL_DEVNET = 901,
+  OPTIMISM_BEDROCK_LOCAL_DEVNET = 902,
```
see if we can change these back without any issues

tynes commented on packages/contracts-bedrock/scripts/Deployer.sol:
```diff
@@ -441,7 +441,7 @@ abstract contract Deployer is Script {
         } else if (chainid == 10) {
             return "optimism-mainnet";
-        } else if (chainid == 900) {
+        } else if (chainid == 900 || chainid == 1337) {
```
don't let this get out of sync with the values in the config

tynes commented on packages/contracts-bedrock/deploy-config/devnetL1.json:
```diff
@@ -1,47 +1,47 @@
 {
-  "l1ChainID": 900,
-  "l2ChainID": 901,
+  "l1ChainID": 901,
```
todo: revert this diff of the chain id changes

tynes commented:
```
    system_config_test.go:103: 
        	Error Trace:	/root/project/op-e2e/actions/system_config_test.go:103
        	Error:      	Not equal: 
        	            	expected: 0x3
        	            	actual  : 0x4
        	Test:       	TestBatcherKeyRotation
        	Messages:   	L2 block with L1 origin that included config change
```

tynes commented on op-e2e/e2eutils/secrets.go:
```diff
@@ -97,8 +97,8 @@ func (m *MnemonicConfig) Secrets() (*Secrets, error) {
 		Deployer:     deployer,
-		CliqueSigner: cliqueSigner,
 		SysCfgOwner:  sysCfgOwner,
+		CliqueSigner: cliqueSigner,
```
reduce unnecessary diff in this file

tynes commented on op-e2e/setup.go:
```diff
@@ -78,74 +79,8 @@ func newTxMgrConfig(l1Addr string, privKey *ecdsa.PrivateKey) txmgr.CLIConfig {
-
-		FinalizationPeriodSeconds: 60 * 60 * 24,
-		MaxSequencerDrift:         10,
-		SequencerWindowSize:       30,
```
double check sequencer window size

tynes commented:
```
        	Error Trace:	/root/project/op-e2e/actions/blocktime_test.go:53
        	            				/root/project/op-e2e/actions/blocktime_test.go:81
        	Error:      	Not equal: 
        	            	expected: 0x11
        	            	actual  : 0x15
        	Test:       	TestBatchInLastPossibleBlocks
```

tynes commented:
For some reason it seems like there are p2p issues between a verifier and sequencer in the tests? Not sure why, still debugging. Asserting that the unsafe block signer is set correctly, seems like the only thing that would cause this

tynes commented on op-e2e/actions/system_config_test.go:
```diff
@@ -92,15 +101,25 @@ func TestBatcherKeyRotation(gt *testing.T) {
-	sysCfg, err := derive.PayloadToSystemConfig(payload, sd.RollupCfg)
-	require.NoError(t, err)
-	require.Equal(t, dp.Addresses.Bob, sysCfg.BatcherAddr, "bob should be batcher now")
+	for i := 0; i <= 24; i++ {
```
Does setting a L2 blocktime of 2 remove the need for this diff?

tynes commented on bedrock-devnet/devnet/genesis.py:
```diff
@@ -8,7 +8,7 @@
 
 GENESIS_TMPL = {
     'config': {
```
Ideally we can remove this template in favor of a command in OP Node that builds the L1 genesis file, then we have 1 source of truth for genesis files which is the op node `genesis` command

tynes commented on packages/contracts-bedrock/deploy-config/devnetL1.json:
```diff
@@ -1,47 +1,47 @@
+  "cliqueSignerAddress": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
+  "l1UseClique": true,
+  "l1StartingBlockTag": "earliest",
+  "l2OutputOracleSubmissionInterval": 6,
```
see if we can increase this safely

tynes replied:
added adee57d17a7708a566108cc0d17ae7dd433f5ea2 to get logging on the rate of output proposals vs L2 blocks. If L2 is running away from L1 it makes the docker devnet setup unusable for dev. Will fix based on the results of the logging

tynes replied:
Can confirm that the devnet will not have issues with an interval of 6 with the current L1 and L2 blocktimes

tynes commented on packages/contracts-bedrock/deploy-config/devnetL1.json:
```diff
@@ -1,47 +1,47 @@
-  "l2BlockTime": 2,
+  "l1ChainID": 901,
+  "l2ChainID": 902,
+  "l2BlockTime": 1,
```
having this as 1 speeds up ci

tynes commented on op-e2e/config/init.go:
```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		panic(err)
+	}
+	slot := common.BigToHash(big.NewInt(int64(entry.Slot)))
```
not ideal that we have to do this, it also means we will have to do it in the python code. ideally we could do it if we were using anvil but it doesn't support the rpc we need. it has an alt state dumping rpc that i'm not exactly sure how to work with to be honest, its binary + compressed, would need to look into it more

tynes replied:
fixed this issue

tynes commented on op-e2e/setup.go:
```diff
@@ -78,81 +80,27 @@ func newTxMgrConfig(l1Addr string, privKey *ecdsa.PrivateKey) txmgr.CLIConfig {
 
-		Premine: make(map[common.Address]*big.Int),
+	// TODO: maybe don't need this?
+	premine := make(map[common.Address]*big.Int)
```
i don't believe we need this anymore

tynes commented on op-node/sources/eth_client.go:
```diff
@@ -403,6 +403,11 @@ func (s *EthClient) GetStorageAt(ctx context.Context, address common.Address, st
 // ReadStorageAt is a convenience method to read a single storage value at the given slot in the given account.
 // The storage slot value is verified against the state-root of the given block if we do not trust the RPC provider, or directly retrieved without proof if we do trust the RPC.
 func (s *EthClient) ReadStorageAt(ctx context.Context, address common.Address, storageSlot common.Hash, blockHash common.Hash) (common.Hash, error) {
+	// TODO: temp
```
not sure why this was an issue, archival mode must be off?

tynes commented on op-e2e/config/init.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	/*
+		L1Allocs, err = genesis.PostProcessL1DeveloperGenesis(L1Allocs, L1Deployments)
```
todo: dead code

tynes committed:
lint: fix

refcell commented on Makefile:
```diff
@@ -77,8 +77,9 @@ devnet-up:
 .PHONY: devnet-up
 
+# alias for devnet-up
 devnet-up-deploy:
-	PYTHONPATH=./bedrock-devnet python3 ./bedrock-devnet/main.py --monorepo-dir=. --deploy
+	PYTHONPATH=./bedrock-devnet python3 ./bedrock-devnet/main.py --monorepo-dir=.
 .PHONY: devnet-up-deploy
```
If `devnet-up-deploy` is now the same as `devnet-up`, can we just remove it or if not point `devnet-up-deploy` to `devnet-up`.

ie.
```suggestion
devnet-up-deploy: devnet-up
```

trianglesphere replied:
+1 to removing `devnet-up-deploy`

tynes replied:
this is now fixed, good call

refcell commented on .circleci/config.yml:
```diff
@@ -886,9 +923,9 @@ jobs:
-            sudo tar -C /usr/local -xzf go1.19.linux-amd64.tar.gz
+            sudo tar -C /usr/local -xzf go1.20.linux-amd64.tar.gz
             export PATH=$PATH:/usr/local/go/bin
             go version
```
go versioning should also be more tightly controlled.
this is error-prone and at times not an obvious toolchain fix.

note, this should just be ticketed and fixed in a separate pr.

tynes replied:
Ok yes I agree

tynes replied:
Reverted this, broke out into its own PR

tynes committed:
docker-compose: fix some security things

tynes committed:
op-node: revert trust rpc setting

tynes committed:
contracts-bedrock: chains constants library

tynes committed:
contracts-bedrock: use Chains.sol

tynes committed:
devnet: revert to using original chainids

tynes committed:
sdk: additional logging

Ensures that the L2 chain isn't running away from
the output proposer. There is a case where the submission
interval is too small such that the L2 chain builds
blocks at a faster rate than the proposer can
currently propose.

tynes committed:
op-bindings: regenerate sourcemaps

tynes committed:
op-bindings: regenerate mips sourcemap

refcell commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -9,22 +9,23 @@
     def __init__(self, **kwds):
         self.__dict__.update(kwds)
 
+
```
nit
```suggestion
```

tynes replied:
the rest of the file uses 2 newlines between things

tynes commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -9,22 +9,23 @@
 parser = argparse.ArgumentParser(description='Bedrock devnet launcher')
 parser.add_argument('--monorepo-dir', help='Directory of the monorepo', default=os.getcwd())
-parser.add_argument('--deploy', help='Whether the contracts should be predeployed or deployed', type=bool, action=argparse.BooleanOptionalAction)
+parser.add_argument('--allocs', help='Path to the allocs file', type=bool, action=argparse.BooleanOptionalAction)
```
This comment is wrong

trianglesphere replied:
Are you planning on fixing it? What should it be instead?

tynes replied:
this has been fixed

refcell commented on packages/sdk/tasks/deposit-erc20.ts:
```diff
@@ -144,30 +145,51 @@ task('deposit-erc20', 'Deposits WETH9 onto L2.')
+      const json = JSON.parse(data.toString())
       contractAddrs = {
-        l1: JSON.parse(data.toString()),
+        l1: {
+          AddressManager: json.AddressManager,
+          L1CrossDomainMessenger: json.L1CrossDomainMessengerProxy,
+          L1StandardBridge: json.L1StandardBridgeProxy,
+          StateCommitmentChain: ethers.constants.AddressZero,
+          CanonicalTransactionChain: ethers.constants.AddressZero,
+          BondManager: ethers.constants.AddressZero,
+          OptimismPortal: json.OptimismPortalProxy,
+          L2OutputOracle: json.L2OutputOracleProxy,
+        },
```
This adds another point of maintenance if we add contracts. Unless there's a good reason, can we revert to passing through the contracts json?

tynes replied:
We cannot pass thru the json directly unfortunately, these tasks will be deprecated along with the sdk soon enough

tynes replied:
The keys changed

refcell commented on packages/sdk/tasks/deposit-eth.ts:
```diff
@@ -88,8 +89,18 @@ task('deposit-eth', 'Deposits ether to L2.')
+      const json = JSON.parse(data.toString())
       contractAddrs = {
-        l1: JSON.parse(data.toString()),
+        l1: {
+          AddressManager: json.AddressManager,
+          L1CrossDomainMessenger: json.L1CrossDomainMessengerProxy,
+          L1StandardBridge: json.L1StandardBridgeProxy,
+          StateCommitmentChain: ethers.constants.AddressZero,
+          CanonicalTransactionChain: ethers.constants.AddressZero,
+          BondManager: ethers.constants.AddressZero,
+          OptimismPortal: json.OptimismPortalProxy,
+          L2OutputOracle: json.L2OutputOracleProxy,
+        },
```
ibid

refcell commented on packages/sdk/tasks/deposit-eth.ts:
```diff
@@ -187,11 +210,11 @@ task('deposit-eth', 'Deposits ether to L2.')
     })
 
-    const opBalanceBefore = await signer.provider.getBalance(
+    const opBalanceBefore = await signer!.provider!.getBalance(
```
I don't think we need the non-null operator here and below since its address is immediately fetched after instantiation. ie, the script will error when calling `signer.getAddress()` on the null `signer = signers[0]`

tynes replied:
My LSP was yelling at me about this

refcell commented:
Current dependencies on/for this PR:
* develop
  * **PR #6412** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6412" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #6552** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6552" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/6412?utm_source=stack-comment).

refcell commented on op-node/cmd/genesis/cmd.go:
```diff
@@ -49,39 +51,36 @@ var Subcommands = cli.Commands{
-			l1Genesis, err := genesis.BuildL1DeveloperGenesis(config)
-			if err != nil {
-				return err
+			if err := config.Check(); err != nil {
+				return fmt.Errorf("deploy config at %s invalid: %w", deployConfig, err)
 			}
```
This can be moved up to immediately after the config is created right?

tynes replied:
It needs to happen after the deployments are set, otherwise it will fail

refcell replied:
Oops missed that, thanks!

refcell reviewed (changes requested):
Small comments and nits mostly. Amazing PR

trianglesphere commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -33,88 +34,112 @@ def main():
 
     os.makedirs(devnet_dir, exist_ok=True)
 
+    if args.allocs:
```
So `--allocs` means building the genesis?

tynes replied:
It technically isn't building the genesis file, its building the allocs field in the genesis file

trianglesphere commented on bedrock-devnet/devnet/__init__.py:
```diff
@@ -33,88 +34,112 @@ def main():
         log.info('Generating L1 genesis.')
-        write_json(paths.genesis_l1_path, GENESIS_TMPL)
+        if os.path.exists(paths.allocs_path) == False:
+            devnet_l1_genesis(paths)
```
what is the difference between `allocs_path` and `genesis_l1_path`? Why do we appear to be generating `allocs_path` with the function `devnet_l1_genesis`?

tynes replied:
The allocs are the result of `debug_dumpBlock` so its the state dump, its called allocs in the genesis struct

trianglesphere commented on op-e2e/README.md:
````diff
@@ -0,0 +1,20 @@
+
+```bash
+make install-geth
+make devnet-allocs
````
Maybe we don't always run `make install-geth`, but we should set the makefile rules up such that `make devnet-allocs` is ran when `devnetL1.json` is newer than anything in `.devnet` & should be ran when running `make test` inside op-e2e. I don't like that we manually have to run several steps to run tests & I think it's likely that we will run into issues where `devnet-allocs` gets out of date.

tynes replied:
I def see what you mean. I should clarify in the readme that you only need to run `make install-geth` once and then again if we update the version

We need to run `make devnet-allocs` if anything in `contracts-bedrock/{src,deploy-config,scripts}` changes

tynes replied:
This doesn't work very well with make. 9ed901f58f15e82480c372b2ca8b726de0fefa37 was an attempt at solving this but it wasn't portable to `sh` and making bash the shell broke ci

trianglesphere reviewed (commented):
First pass seems solid, but there's a lot to review.

tynes commented:
A point of confusion is the difference between allocs and L1 genesis. Right now the implementation separates the creation of generating the L1 state and the L1 genesis block itself. It was slightly easier to implement it this way but I do believe that I could refactor things so that the L1 genesis is created and there isn't a concept of allocs if that is preferred. The L1 genesis has a timetamp on it which could go stale if its not created dynamically, but that is something that we could always modify at time of loading the file

tynes committed:
revert: ci go1.20

In favor of https://github.com/ethereum-optimism/optimism/pull/6557

tynes committed:
genesis: add clarifying comment

tynes committed:
bedrock-devnet: fix help comment

tynes committed:
makefile: cleanup

tynes committed:
makefile: ensure up to date

tynes committed:
makefile: use bash

tynes committed:
Revert "makefile: use bash"

This reverts commit 6f364edbb3502a89a6df2fc4dc2056b2640426ed.

tynes committed:
Revert "makefile: ensure up to date"

This reverts commit 08a230d02b6f0fd28efb0994536b989187ba01ea.

tynes commented:
Can we follow up in a future PR for the auto detect/rebuild functionality? cc @trianglesphere

protolambda commented on op-chain-ops/genesis/config.go:
```diff
@@ -74,10 +77,15 @@ type DeployConfig struct {
 	// L2OutputOracleChallenger is the address of the account that challenges L2 outputs.
 	L2OutputOracleChallenger common.Address `json:"l2OutputOracleChallenger"`
 
+	// CliqueSignerAddress represents the signer address for the clique consensus engine.
+	// It is used in the multi-process devnet to sign blocks.
+	CliqueSignerAddress common.Address `json:"cliqueSignerAddress"`
+	// L1UseClique represents whether or not to use the clique consensus engine.
+	L1UseClique bool `json:"l1UseClique"`
+
```
Fine to make it optional in this PR, but would like to get rid of clique soon, since it's not going to be around for very long upstream.

tynes replied:
I definitely would like to get rid of clique too, the problem is that the devnet still relies on it, I made a linear issue on this. https://linear.app/optimism/issue/CLI-4309/update-l1-node-in-docker-compose-devnet

protolambda commented on op-e2e/e2eutils/setup_test.go:
```diff
@@ -22,6 +23,7 @@ func TestSetup(t *testing.T) {
 		MaxSequencerDrift:   40,
 		SequencerWindowSize: 120,
 		ChannelTimeout:      120,
+		L1BlockTime:         15,
```
Why 15 and not 12? I think it previously defaulted to 12.

tynes replied:
I thought that I made this exactly what the value was previously on purpose, maybe I made a mistake. The test works either way

protolambda commented on op-e2e/actions/l2_verifier_test.go:
```diff
@@ -34,6 +34,7 @@ func TestL2Verifier_SequenceWindow(gt *testing.T) {
 		MaxSequencerDrift:   10,
 		SequencerWindowSize: 24,
 		ChannelTimeout:      10,
+		L1BlockTime:         15,
```
Same here, why 15 and not 12?

tynes replied:
The default is 15 so I am going to leave as is

https://github.com/ethereum-optimism/optimism/blob/00e7095ce39a60f74de27189c031c384405a4f95/op-e2e/actions/l1_replica_test.go#L21-L26

protolambda reviewed (approved):
generally looks good

refcell reviewed (approved):
Looks good to me!

tynes committed:
makefile: automate rebuild

tynes committed:
makefile: fix
